### DISCUSSION TITLE
Whats the best practice for a browser incompatibility message?

### DISCUSSION THREAD

**Discussion by maerzhase** [vercel/next.js#16782] [Help]
Hej folks! 
Since I actually could not find any good examples or articles to read on the web I wanted to ask in the community: 
__What do you consider a best practise for displaying a "Browser Incompatibility Message" within a next.js app?__

I have a few ideas on __where__ (in the server, in a custom app, on each page individual, etc.) and __how__ (with a middleware or decorator that parses the UserAgent and does something depending on it) to build a solution for it but I am hesitating to implement one because I was surprised not to find a common example on how someone would do that within next.js. Before I start building a custom solution for it I wanted to make sure that I am not missing something totally obvious.

Just assume that you build a project with a certain compatibility in mind e.g. "only Chrome >=v75 and Firefox >=v77" and all the other browsers should get a designated message displayed that informs the user about the browser he can use to visit the website. 

Ideally the solution would work without a custom server and supports even outdated browsers, e.g. ie11. 

Happy to hear all your ideas and thoughts!

**Comment by danielkcz** [comment]
@maerzhase I am curious about the approach you have settled within the end?

**Reply by maerzhase** [reply]
My current approach is parsing the user-agent from the request in the `getInitialProps` function of the `_app` and do a redirect based on that. I am using `bowser` for user-agent parsing and `isbot` for checking for bots

```es6
import { getParser } from 'bowser';
import isBot from 'isbot';

const BROWSER_SATISFACTION = {
  Chrome: '>=52',
  Firefox: '>=50',
  'Internet Explorer': '>=11',
  Safari: '>=5',
};

const browserIsSupported = useragent => {
  const browser = getParser(useragent);
  const isSupported = browser.satisfies(BROWSER_SATISFACTION);
  if (typeof isSupported === 'undefined') {
    console.log('[BC] is unknown ');
    return true;
  } else if (isSupported) {
    console.log('[BC] is supported');
    return true;
  } else {
    console.log('[BC] not supported');
    return false;
  }
};

export const handleRedirectToNotSupportedPage = (req, res) => {
  const useragent = req.headers['user-agent'];
  console.log('[BC] checking user agent');
  console.log('[BC]', useragent);

  // if the useragent is a crawler/bot let him through
  if (isBot(useragent)) {
    console.log('[BC] isBot');
    return;
  }
  // check for all supported browsers; if one doesnt match
  // we redirect to the redirectPath;
  if (browserIsSupported(useragent)) {
    return;
  }
  res.writeHead(302, {
    Location: '/not-supported'
  });
  res.end();
  return;
};
```

To me that was the straight-forward solution but I would like to see others. I am super surprised that nobody replied to this (imo) common use-case.

**Reply by danielkcz** [reply]
Thanks a lot. I am not sure if it's that common problem. I mean most apps probably choose the approach of polyfills or not using ES2015+ stuff instead of dealing with browser support.

I am still pretty much new to Next. I am not sure what is the proper way in V10+ where `getInitialProps` is not recommended anymore. I guess it's `getServerSideProps`, but it feels weird to essentially opt for server rendering just because of this tiny thing.

I was attempting a purely client-side solution that can first run this detection and then decide if continue normally or redirect to eg. `/not-supported` as in your case. Unfortunately, in IE11 I would have to polyfill a whole bunch of stuff first so this detection can even run. That's far from ideal.

**Reply by maerzhase** [reply]
Yes you are right `getInitialProps` is not recommended anymore but same principle would be possible with `getServerSideProps`. 

I would need to fully wrap my head around it, but as you mention: In a pure client-side solution you would still need to ensure that you app can be rendered on any browser and then do redirect.

My preferred solution uses the server because it could potentially even serve a static `.html` file, that is completely independent of the whole application, to "users" that are not supported.

IMO it should be a common use-case because its better to serve a designated page to outdated browser rather serving a semi optimised and polyfilled version of the content... :D
